### PR TITLE
Wallet

### PR DESCRIPTION
**Pull request by Yostra**
- Keeps all coin records and unspent_for_wallet records in memory. Operations that modify db, modify the cache as well. In case of failure to write into db, cache gets fully rebuilt. 
- Same ^ for TXs. Keeps all tx records and unconfirmed_per_wallet records in memory...
- Keeps last few new peak messages while wallet is syncing and reapplies them after sync is done. 
- Fixes TX reorgs. 
- Full node & wallet were not reseting spent height back to 0 in case of rollback.
- Wallet to call rollback only when new peak.prev_hash is different from current_peak.header_hash
- Use get_block_record_from_db, block_recrod cache is limited in size. 
-  TX block used for generator ref can be reorged, handle that case.

### CODE REVIEW

File: chia/wallet/wallet_coin_store.py [hunk 1]
```diff
@@ -18,9 +17,10 @@ class WalletCoinStore:
     """
 
     db_connection: aiosqlite.Connection
+    # coin_record_cache keeps ALL coin records in memory. [record_name: record]
```
Comment (mariano54) [inline chia/wallet/wallet_coin_store.py:20]: This is fine for wallet, but not great for full node. I ran some simulations and at height 500k it slowed down significantly with all coins in memory, FYI

Comment (arvidn) [reply chia/wallet/wallet_coin_store.py:20]: is it important to keep these in memory? I would expect sqlite to have its own cache, and that lookups for the same coins would be quick. In fact, I would trust sqlite's cache and the page cache more than our python collections.

File: chia/wallet/wallet_coin_store.py [hunk 2]
```diff
@@ -62,58 +62,65 @@ async def create(cls, wrapper: DBWrapper):
...
         return self
 
     async def _clear_database(self):
         cursor = await self.db_connection.execute("DELETE FROM coin_record")
         await cursor.close()
         await self.db_connection.commit()
 
+    async def rebuild_wallet_cache(self):
+        # First update all coins that were reorged, then re-add coin_records
+        all_coins = await self.get_all_coins()
```
Comment (arvidn) [inline chia/wallet/wallet_coin_store.py:77]: this returns all coins in this wallet, including spent ones, right?
It's not clear that this will fit in memory, especially not when parsed into python types.

File: chia/wallet/wallet_coin_store.py [hunk 3]
```diff
@@ -62,58 +62,65 @@ async def create(cls, wrapper: DBWrapper):
...
+        )
+        await cursor.close()
 
     # Update coin_record to be spent in DB
-    async def set_spent(self, coin_name: bytes32, height: uint32):
+    async def set_spent(self, coin_name: bytes32, height: uint32) -> WalletCoinRecord:
         current: Optional[WalletCoinRecord] = await self.get_coin_record(coin_name)
-        if current is None:
-            return None
+        assert current is not None
```
Comment (mariano54) [inline chia/wallet/wallet_coin_store.py:123]: We should assert that its unspent here, as we do in the full node

File: chia/wallet/wallet_coin_store.py [hunk 4]
```diff
@@ -62,58 +62,65 @@ async def create(cls, wrapper: DBWrapper):
...
         cursor = await self.db_connection.execute("DELETE FROM coin_record")
         await cursor.close()
         await self.db_connection.commit()
 
+    async def rebuild_wallet_cache(self):
+        # First update all coins that were reorged, then re-add coin_records
+        all_coins = await self.get_all_coins()
+        self.unspent_coin_wallet_cache = {}
+        self.coin_record_cache = {}
+        for coin_record in all_coins:
```
Comment (arvidn) [inline chia/wallet/wallet_coin_store.py:80]: If we do need to iterate over all coins, I think this would be a lot more efficient to iterate over the sqlite cursor here. To stream them in from the database.

File: chia/wallet/wallet_state_manager.py [hunk 5]
```diff
@@ -655,7 +661,7 @@ async def coin_added(
         wallet_type: WalletType,
         height: uint32,
         all_outgoing_transaction_records: List[TransactionRecord],
-    ):
+    ) -> WalletCoinRecord:
```
Comment (arvidn) [inline chia/wallet/wallet_state_manager.py:664]: I don't see the call to this function that use the new return value. Is this change necessary?

File: chia/wallet/wallet_state_manager.py [hunk 6]
```diff
@@ -799,9 +806,10 @@ async def get_filter_additions_removals(
             fork_h: int = fork_point_with_peak
         elif new_block.prev_header_hash != self.constants.GENESIS_CHALLENGE and self.peak is not None:
             # TODO: handle returning of -1
+            block_record = await self.blockchain.get_block_record_from_db(self.peak.header_hash)
```
Comment (arvidn) [inline chia/wallet/wallet_state_manager.py:809]: is this addressing the case where `fork_h` can be -1?
If so, please add an `assert fork_h >=0` below, and remove the comment above.

File: chia/wallet/wallet_transaction_store.py [hunk 7]
```diff
@@ -19,17 +19,14 @@ class WalletTransactionStore:
...
-    tx_wallet_cache: Dict[int, Dict[Any, Set[bytes32]]]
     tx_submitted: Dict[bytes32, Tuple[int, int]]  # tx_id: [time submitted: count]
     unconfirmed_for_wallet: Dict[int, Dict[bytes32, TransactionRecord]]
 
     @classmethod
-    async def create(cls, db_wrapper: DBWrapper, cache_size: uint32 = uint32(600000)):
+    async def create(cls, db_wrapper: DBWrapper):
         self = cls()
 
-        self.cache_size = cache_size
```
Comment (arvidn) [inline chia/wallet/wallet_transaction_store.py:32]: is this part of the patch effectively making the cache unlimited?

File: chia/wallet/wallet_transaction_store.py [hunk 8]
```diff
@@ -225,30 +222,26 @@ async def increment_sent(
...
-    async def tx_reorged(self, tx_id: bytes32):
+    async def tx_reorged(self, record: TransactionRecord):
         """
         Updates transaction sent count to 0 and resets confirmation data
         """
-
-        current: Optional[TransactionRecord] = await self.get_transaction_record(tx_id)
-        if current is None:
-            return None
         tx: TransactionRecord = TransactionRecord(
```
Comment (arvidn) [inline chia/wallet/wallet_transaction_store.py:229]: this looks like a candidate for using
```
tx: TransactionRecord = dataclasses.replace(
    record, confirmed_at_height=uint32(0),
    confirmed=False,
    sent=uint32(0),
    sent_to=[],
    trade_id=None)
```

**Review by arvidn — COMMENTED** [review]
I think we need unit tests for the primitives, like `coin_store` as well. Not just higher level tests that covers a specific scenario.

File: chia/wallet/wallet_coin_store.py [hunk 9]
```diff
@@ -62,58 +62,65 @@ async def create(cls, wrapper: DBWrapper):
...
+            if record.spent and name in self.unspent_coin_wallet_cache[record.wallet_id]:
+                self.unspent_coin_wallet_cache[record.wallet_id].pop(name)
+            if not record.spent:
+                self.unspent_coin_wallet_cache[record.wallet_id][name] = record
+        else:
+            if not record.spent:
+                self.unspent_coin_wallet_cache[record.wallet_id] = {}
+                self.unspent_coin_wallet_cache[record.wallet_id][name] = record
 
-        await self.wallet_cache_lock.acquire()
```
Comment (mariano54) [inline chia/wallet/wallet_coin_store.py:83]: Is it ok to remove this lock?

File: chia/wallet/wallet_coin_store.py (hunk 2 above)
Comment (mariano54) [reply chia/wallet/wallet_coin_store.py:77]: This is only coins that are owned by this wallet, so its much less than the whole blockchain

File: chia/full_node/full_node.py [hunk 10]
```diff
@@ -1153,7 +1153,12 @@ async def respond_block(
                 )
                 assert result_to_validate.required_iters == pre_validation_results[0].required_iters
                 added, error_code, fork_height = await self.blockchain.receive_block(block, result_to_validate, None)
-
+                if (
+                    self.full_node_store.previous_generator is not None
+                    and fork_height is not None
+                    and fork_height < self.full_node_store.previous_generator.block_height
+                ):
+                    self.full_node_store.previous_generator = None
```
[the remark this answers is not in the document]
Comment (mariano54) [reply chia/full_node/full_node.py:1161]: Yes it's only used for farming a block. :+1: for test. We can just do a reorg in the first test in test_full_node.py

File: chia/wallet/wallet_coin_store.py [hunk 11]
```diff
@@ -62,58 +62,65 @@ async def create(cls, wrapper: DBWrapper):
...
 
+    async def rebuild_wallet_cache(self):
+        # First update all coins that were reorged, then re-add coin_records
+        all_coins = await self.get_all_coins()
+        self.unspent_coin_wallet_cache = {}
+        self.coin_record_cache = {}
+        for coin_record in all_coins:
+            name = coin_record.name()
+            self.coin_record_cache[name] = coin_record
+            if coin_record.spent is False:
```
Comment (Nikolaj-K) [inline chia/wallet/wallet_coin_store.py:83]: `not`?

File: chia/wallet/wallet_coin_store.py (hunk 1 above)
Comment (Yostra) [reply chia/wallet/wallet_coin_store.py:20]: @mariano54 Yes, this is only for wallet. 
@arvidn Yes, performance. SQlite memory cache is limited in configurability, max size of memory is not per table configurable. Header blocks would dominate the coin records.

Comment (Yostra) [reply chia/wallet/wallet_coin_store.py:20]: Also the coin set (even for whales) is not that big.

File: chia/full_node/coin_store.py [hunk 12]
```diff
@@ -202,7 +202,7 @@ async def rollback_to_block(self, block_index: int):
                 new_record = CoinRecord(
                     coin_record.coin,
                     coin_record.confirmed_block_index,
-                    coin_record.spent_block_index,
```
[the remark this answers is not in the document]
Comment (Yostra) [reply chia/full_node/coin_store.py:205]: Consistency between what is on disk and what is in memory. There wasn't specific issue that was caused by this bug as queries that used this field always read from the disk.

File: chia/wallet/wallet_node.py [hunk 13]
```diff
@@ -499,6 +502,7 @@ async def sync_job(self) -> None:
                 break
             asyncio.create_task(self.check_new_peak())
             await self.sync_event.wait()
+            self.last_new_peak_messages = LRUCache(10)
```
[the remark this answers is not in the document]
Comment (Yostra) [reply chia/wallet/wallet_node.py:505]: not deliberate, I'll make a change. Yup, these messages are ignored while syncing. Replaying last few right after long sync is done will kick start the sync again. (If chain advanced while we were syncing)

File: chia/wallet/wallet_node.py [hunk 14]
```diff
@@ -513,6 +517,8 @@ async def sync_job(self) -> None:
             finally:
                 if self.wallet_state_manager is not None:
                     self.wallet_state_manager.set_sync_mode(False)
+                for peer, peak in self.last_new_peak_messages.cache.items():
+                    asyncio.create_task(self.new_peak_wallet(peak, peer))
```
[the remark this answers is not in the document]
Comment (Yostra) [reply chia/wallet/wallet_node.py:521]: new_peak_lock will ensure that these tasks execute one by one.

File: chia/wallet/wallet_state_manager.py (hunk 5 above)
Comment (Yostra) [reply chia/wallet/wallet_state_manager.py:664]: Not necessary, I started in direction of rebuilding the cache by tracking of the writes that failed first but then changed direction to just rebuild from scratch. This is artifact, I'll remove this.

File: chia/wallet/wallet_transaction_store.py (hunk 7 above)
Comment (Yostra) [reply chia/wallet/wallet_transaction_store.py:32]: yes

File: chia/wallet/wallet_coin_store.py (hunk 9 above)
Comment (Yostra) [reply chia/wallet/wallet_coin_store.py:83]: yes, the cache only gets changed in other lock. (block)

File: chia/wallet/wallet_coin_store.py (hunk 2 above)
Comment (Yostra) [reply chia/wallet/wallet_coin_store.py:77]: Yes, this is for single wallet. Not concerned about memory size here.

File: chia/wallet/wallet_state_manager.py (hunk 6 above)
Comment (Yostra) [reply chia/wallet/wallet_state_manager.py:809]: changed the arg `self.blockchain.block_record(self.peak.header_hash)` as block_record cache is limited in size, so there's no really guaranty self.peak.header_hash is in it.

Comment (Yostra) [reply chia/wallet/wallet_state_manager.py:809]: We should probably avoid using self.blockchain.block_record everywhere

Comment (Yostra) [reply chia/wallet/wallet_state_manager.py:809]: (not related ot -1)

**Review by mariano54 — DISMISSED** [review]
I think we should merge this as it fixes some key issues with stuck pending transactions and reorgs. It's also not  high risk because it's mainly wallet.
